### DISCUSSION TITLE
Alternate Constructor Syntax

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#3346] [Language Ideas]
I had an idea last night that I wanted to toss out there.  This is related to some of the discussion I've seen about records.

Today, I can have a class like this:
````
public class Foo {
  public string Name {get; set;}
  public int Age {get; set; }

  public Foo(){

  }

  public foo(string name, int age){
    this.Name = name;
    this.Age = age;
  }
}
````

There are four main ways that the type may be constructed:
````
var A = new Foo();
var B = new Foo("John", 21);
var C = new Foo(name: "John", age: 21);
var D = new Foo(){Name = "John", Age = 21};
var E = new Foo {Name = "John", Age = 21};
````

Currently ````E```` means "Make a new Foo using the default constructor.  Then set some properties".

What if the meaning of ````E```` were to be changed to:
Find a constructor such that all parameters are satisfied by the { } arguments and any leftover arguments map to publicly settable properties.

**Comment by HaloFour** [comment]
Constructor parameters and properties have different naming conventions, and the team had already decided to not try to map properties to constructors based on a case insensitive or other transformation of the names.

**Comment by Unknown6656** [comment]
> What if the meaning of `E` were to be changed to:
> Find a constructor such that all parameters are satisfied by the { } arguments and any leftover arguments map to publicly settable properties.

Then this would be a breaking change and therefore never be accepted as a proposal.

**Comment by CyrusNajmabadi** [comment]
> What if the meaning of E were to be changed to:

We will not change the meaning of code without staggering benefit.  The risk of breaks is just not worth it.
